### ISSUE TITLE
Tracking Issue for TCP_QUICKACK support on Linux

### ISSUE THREAD

**Issue by berendjan** [rust-lang/rust#96256]
Feature gate: `#![feature(tcp_quickack)]`

This is a tracking issue for linux specific socket option `TCP_QUICKACK`

from linux man page TCP:
        _"Enable quickack mode if set or disable quickack mode if
        cleared.  In quickack mode, acks are sent immediately,
        rather than delayed if needed in accordance to normal TCP
        operation.  This flag is not permanent, it only enables a
        switch to or from quickack mode.  Subsequent operation of
        the TCP protocol will once again enter/leave quickack mode
        depending on internal protocol processing and factors such
        as delayed ack timeouts occurring and data transfer.  This
        option should not be used in code intended to be portable."_

### Public API
```rust
// std::os::linux::net

// sealed trait, implemented for std::net::TcpStream
pub trait TcpStreamExt: Sealed{
  fn quickack(&self) -> io::Result<bool>;
  fn set_quickack(&self, quickack: bool) -> io::Result<()>;
}
```

### Steps / History

- [x] Implementation: 
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

[Zulip Discussion](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/First.20contribution.20TCP_QUICKACK)
[Windows .NET Discussion](https://github.com/dotnet/runtime/issues/798)

Seems the windows API is significantly different from the Linux one to warrant a Linux specific trait.

**Comment by jmillikin** [comment]
FYI: The same modules are also being touched as part of the work on supporting abstract socket addresses -- see https://github.com/rust-lang/rust/issues/85410 and https://github.com/rust-lang/rust/pull/101967 for details.

I tried to keep the code intact (merely relocated), so hopefully there won't be any major conflicts.

**Comment by joshtriplett** [comment]
Please consider adding a comment to `set_nodelay` that references this (only on Linux), and a comment to `set_quickack` that references `set_nodelay` and the interaction between quickack and nodelay (namely, that you should usually disable one or the other, because having both enabled causes delays).

Please also consider implementing support on Windows, which uses `SIO_TCP_SET_ACK_FREQUENCY`.

**Comment by ChrisDenton** [comment]
> Please also consider implementing support on Windows, which uses `SIO_TCP_SET_ACK_FREQUENCY`.

I think that would be a separate issue? It's not an exactly equivalent API and in any case obviously wouldn't be in `std::os::linux`.

**Comment by joshtriplett** [comment]
Yes, absolutely. I mentioned it here because it's related, but it's certainly not a blocker.
